### ISSUE TITLE
Helios modified send keys are internally also recognized as un-modified send keys.

### ISSUE THREAD

**Issue by Shaftoe62** [HeliosVirtualCockpit/Helios#793]
Hi,

I ran into an issue where send keys commands for function keys (F1, F2, etc)  with modifiers also trigger the same function keys without the modifiers.

For example, if you have a panel that has it's visibility toggled or set by F10 on the keyboard (in sync with the DCS map view), that will also happen if you have a Helios button that outputs {RALT}{RSHIFT}{F10}.

Helios will internally recognize that as a single F10 keystroke as well and will therefore toggle or set the visibility of that panel.

I posted a workaround solution for this on Discord but I thought I mention it here as well because it shouldn't happen in the first place.

If I send a modified key I don't expect it to also trigger events for the un-modified key, right?

But that won't be an easy fix I guess...

**Comment by Shaftoe62** [comment]
@BlueFinBima

Other software: No

But maybe I was not clear enough:


I have a lot of Helios buttons that use the keyboard emulator so I can use the same keyboard combo's in - for example - a StreamDeck or Voice attack or other HIDs.

So for the MFD action buttons I use all function keys with modifiers like {LALT}{LSHIFT}{{F10} for the Left MFD L4 button and {RALT}{RSHIFT}{{F10} for the Right MFD L4 button.

That all works perfectly for about a year now.

But now I wanted to use the F10 key to set the visibility of the NOAA compass to true to make it show up when I go to the DCS map view. And at the same time set the visibility of the other overlays to false.

And F1 does the reverse.

To achieve that I first made a binding for those panels to the mentioned (unmodified) F-keys.

But then I noticed that the visibility also changed when I pushed the MFD L4 button.

So, that's why I said that Helios is internally also responding to modified key sends when only an unmodified key is bound to a specific action.

The way you describe your debugging test makes it sound as if that is intentional. If a modified key is supposed to do what you describe, any action that is configured to be triggerd by just the unmodified key will indeed be triggered.

Just to make sure: are we talking about the same thing here? I'm talking about the Keyboard send keys binding. Not the Keyboard press key/release key bindings.

The only way you can reproduce this behaviour is to check it in a working profile, I think. For that you need to set it up as I described.

In any case, whether this behaviour is intentional or not, it means that as soon as you have used any key in any modified form, you cannot use that key in it's unmodified form for any other action/trigger anymore. 

At least not as long as Helios itself does not (internally) check if a key is send in its unmodified form or not.

And like I said before, I don't expect this to be solved (any time soon) in Helios. That's why I posted my solution on Discord.

If you take a good look at that you will understand what Helios is currently not doing but what it should be doing.

If you don't have the time to make a quick profile to reproduce this behaviour but, on the other hand, are willing to look into this matter, I can make the profile for you. Just let me know.

**Comment by Shaftoe62** [comment]
@BlueFinBima 

By the way, the "Function Key Filter" demonstrates the usefulness of the Wires quite nicely. I'm really happy to have taken the effort to look into these.

It could perhaps be done without wires as well but the wires make it a lot better "debuggable".

**Comment by BlueFinBima** [comment]
Thanks for the additional information - I didn't realise that you were referring to an keyboard input bindings.  This is working as designed (although this may not be as you desire it to be).  It looks like the original intent of the code was to ignore modifiers, and then `LeftControl LeftShift` was added as a special case.  It is not easy to run a debugger on this DirectXInput code which would make changing it a little challenging.
![image](https://github.com/HeliosVirtualCockpit/Helios/assets/18526232/8e902642-6a75-4051-ab07-2a534563ac21)

I'm going to flag this as an enhancement request but it will go into the fridge until someone with enough time and interest makes themselves known and performs the work.

Thanks for raising this.

**Comment by Shaftoe62** [comment]
Indeed, the existence of those pre-defined LeftControl LeftShift combo's suggest that they have been part of an effort from the past to circumvent the issue in some way.

I'm pretty sure that if you dive into the original code deep enough you will find that these specific combo's are being filtered out in the way I am proposing.

I guess they are part of an unfinished solution to the issue.

And I'm sorry to say that my coding knowledge will not be good enough to help in any way. Nor do I have the time (or stomach) to set up a dev environment for it. In C++ I suppose?

But again, the Function Key Filter I posted on Discord does the job pretty well anyway. So If anybody shows up with a (support) question about this, you know there is a workaround. Other than not using keys both modified and unmodified, that is.

There is one more thing I will do as a test to wrap this up. And that is to check if my hunch that this issue will not occur with the predefined LeftControl LeftShift combo's is correct. That will prove that part of the solution is already build in somewhere but, for some reason, was not completed to work generically for all modifier combo's.

I will let you know when I have some results on that.

Thanks a lot for indulging me on this issue.

**Comment by Shaftoe62** [comment]
@BlueFinBima

P.S. For me it was not an option to switch to using only the predefined LeftControl LeftShift combo's.

That would limit the available amount of combo's way to much and I allready have all those combo's integrated in my VKB controllers, StreamDecks and VoiceAttack. it would be way to much work to go and change all that.

So, that motivated me to find another solution.

Did I ever tell you about the database I setup to keep track of all those key combo's for all my HID's, Helios and Voice Attack for the DCS profiles of all the modules I own?

At some point it got so bad to think of new and unused combo's that I made a function to do that for me by generating all possible iterations of key combo's and giving me the ones that were not allready in the database. In a specific order and whether there was a left and right variant still available.

The AH-64D alone already uses up about 700 combo's for each role (Pilot/CPG). 

Most of which are shared by the way.

Actually that is what enabled me to build a Helios profile that would work for both roles. By configuring both the DCS profiles to completely work with (the same) key combo's and then make all the buttons in Helios send those combo's.

And don't think I'd be typing in all those combo's into the database manually. I figured out a way to use the DCS option to export a profile as HTML to get them into the database.

How nerdy can you get?

**Comment by Shaftoe62** [comment]
@BlueFinBima

Hi,

It is as I suspected. The LeftControl LeftShift combo's are specifically excluded from triggering a single keystroke as opposed to any other modifier combo.

But that is actually what it says on the box. If you select a keyboard binding to trigger an action you can choose from: "Any Modifiers" or "LeftControl LeftShift".

![image](https://github.com/HeliosVirtualCockpit/Helios/assets/53898212/4f5faa25-48d5-4ba1-aa5c-2dd6751ff8c8)

So it is correct as it is right now. It's just that I had never before used a "Any Modifiers" binding without any modifiers that had me confused.

But it would be an enhancement nonetheless if you could use an isolated keystroke without the need for a filter like I made to prevent the issue.

Perhaps it's an idea, and less complicated to implement, to introduce a third category:

Single Keystroke
Any Modifiers
LeftControl LeftShift

But I think that would actually make the LeftControl LeftShift option redundant...


I will also include a basic profile I used to test the behaviour.


[Shaftoe62-modified-keys.zip](https://github.com/HeliosVirtualCockpit/Helios/files/13988888/Shaftoe62-modified-keys.zip)
